docs/ocean/design_docs/vert-lagrang-remap/vert-lagrang-remap.rst

```diff
@@ -0,0 +1,544 @@
+
+   \frac{\partial (h_k \phi_k)}{\partial t} + \nabla \cdot (h_k \mathbf{u}_k \phi_k) = [D_h^{\phi}]_k + [D_v^{\phi}]_k + F_k^{\phi}
+   
+The time-stepping algorithm (RK4 or split-explicit) yields the updated 
+variables :math:`u_k^{lg},h_k^{lg},\phi_k^{lg}`, where the superscript
+*lg* is used to designate the values after the Lagrangian step.
+
+Note that the vertical mixing terms :math:`D_v^h, D_v^{\phi}` 
+are retained here. We opt to compute these terms prior to remapping as this 
+allow for future development in which the dynamics are subcycled relative to 
+the thermodynamics and remapping is scheduled on the thermodynamic timestep. 
+This computation of vertical mixing terms prior to remapping is similar to 
+both MOM6 and HYCOM. We anticipate that there could be a trade-off between (a)
+loss of accuracy of vertical mixing terms when their computation precedes 
+remapping due to grid deformation and (b) loss of accuracy when their 
+computation follows remapping due to remapping errors in vertical gradients of 
+prognostic variables. We do not intend to test this at this time.
+
+The target grid needs to be determined after the solution for prognostic 
+variables so that the vertical Lagrangian-remapping method is general enough to
+be used with coordinate systems that depend on the ocean state (this includes 
+the z-star coordinate system in which SSH perturbations are vertically 
+distributed between layers). We do not present an algorithmic design for 
+regridding to coordinate systems not already supported in MPAS-Ocean, as this 
+will be the subject of future development. For now, the target grid is based on a 
+constant set of z-levels that are specified at initialization.
+
+For the regridding step, layer thicknesses are set according to the target 
+grid, conserving volume:
+
+.. math::
+
+   h_k^{t+1} = h_k^{target}
+   
+   \sum_{k=1}^{kmax}h_k^{t+1} = \sum_{k=1}^{kmax}h_k^{lg}
+
+
+For the remapping step, velocities (edge-normal) and scalars are remapped to 
+the target grid, conserving volume flux and scalar concentration:
+
+.. math::
+
+   \sum_{k=1}^{kmax} u_k^{t+1} h_k^{t+1} = \sum_{k=1}^{kmax} u_k^{lg} h_k^{lg}
+   
+   \sum_{k=1}^{kmax} \phi_k^{t+1} h_k^{t+1} = \sum_{k=1}^{kmax} \phi_k^{lg} h_k^{lg}
+
+The vertical velocity across layer interfaces may be computed anytime after 
+regridding. It can be computed as 
+
+.. math::
+
+   w = - \nabla \cdot (h_k \mathbf{u}_k) - (h_k^{t+1} - h_k^t)/dt
+
+or
+
+.. math::
+
+   w = (h_k^{t+1} - h_k^{lg})/dt
+
+The choice between the two is discussed in the Implementation section.
+
+
+Implementation
+--------------
+
+Implementation: Ability to perform vertical Lagrangian-remapping
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2020/12/15
+
+Contributors: Darren Engwirda, Xylar Asay-Davis, Carolyn Begeman
+
+Namelist options:
+
+- To turn VLR on/off: 
+  :code:`ALE_vertical_grid, config_vert_lagrangian_remap = .true. or .false.`
```

Comment:
I see this as a choice, so I would prefer being more explicit about `false`:
`config_tracer_vert_advection_method = 'transport' or 'remap'`
Then in the code, `if (config_tracer_vert_advection_method.eq.'transport')` is more clear.

The problem is that this list is now confusing:
```
&advection
    config_vert_tracer_adv = 'stencil'
    config_vert_tracer_adv_order = 3
    config_horiz_tracer_adv_order = 3
    config_coef_3rd_order = 0.25
    config_monotonic = .true.
/
```
because the first two flags look like they apply to remapping. I see the first is not even used. So I would do this:
```
&advection
    config_vert_tracer_advection_method = 'transport' or 'remap' (new)
    config_vert_tracer_adv = 'stencil' --- delete
    config_vert_tracer_adv_order = 3 --> config_vert_tracer_transport_order = 3
    config_vert_tracer_remap_order = 3 (new)
    config_horiz_tracer_adv_order = 3
    config_coef_3rd_order = 0.25
    config_monotonic = .true.
/
```
I think the new flags belong in the advection list, not the ALE list.

Reply:
@mark-petersen I see. That seems appropriate to me.

Reply:
I think we may actually need the single option to toggle lagrangian-remap on / off, since this'll be changing the vertical advection scheme / terms for tracers, thickness and momentum (rather than just tracers) and it'll be necessary for everything to be done either as `remap` or `transport` consistently.

No problem to opt for more explicitness in the options if that's desired, though `config_vert_tracer_adv_order = {1,2,3,5}` would definitely make direct sense (to me) in the `remap` case too.

Reply:
@dengwirda I see what you mean about thickness and momentum. We could instead use
```
config_vert_advection_method = 'transport' or 'remap'
```
for tracers, thickness and momentum.  

My point was just that if you use a Boolean, it does not communicate what happens for `false`. A Boolean is appropriate if it is really on/off, but here false implies that you turn on transport instead.

Reply:
> I think we may actually need the single option to toggle lagrangian-remap on / off,since this'll be changing the vertical advection scheme / terms for tracers, thickness and momentum (rather than just tracers)

Is this an appropriate fix, make some of the remap-related names more general:
```
&advection
    config_vert_tracer_advection_method = 'transport' or 'remap' --> config_vert_advection_method = 'transport' or 'remap'
    config_vert_tracer_transport_order = 3
    config_vert_tracer_remap_order = 3 --> config_vert_remap_order = 3
    config_horiz_tracer_adv_order = 3
    config_coef_3rd_order = 0.25
    config_monotonic = .true.
```
and should we add all of the remap options here:
```
   config_vert_remap_method = 'pqm'
   config_vert_remap_edge_interp_method = `p5e`
   etc.
```

Reply:
@cbegeman, yes I think `config_vert_advection_method =` `transport` or `remap` makes sense, although what does everyone think about using `flux-form` in place of `transport`?

In terms of the remap options, we may be able to get away with simplifying a little, perhaps something like:
```
(method = pcm) or (order = 1) --> pcm_method
(method = plm) or (order = 2) --> plm_method
(method = ppm) or (order = 3) --> ppm_method + p3e_method
(method = pqm) or (order = 5) --> pqm_method + p5e_method
```

It is possible to mix and match some of these options in `PPR`, but I think I'd expect the combinations above to be the ones typically used in practice.

Reply:
@dengwirda `flux-form` sounds clear to me.

I'm in favor of simplifying. Are you amenable to cutting `config_vert_remap_method` and only using `config_vert_remap_order` since we don't expect folks to mix and match?

If not, do you expect that logic so be applied sequentially so that if order is higher than method, the method is changed; and if method is higher than order, the method stays the same? Given this logic it would be pretty hard to mix and match anyhow, right?

Reply:
Oh, I just meant that I think the edge interp scheme can be inferred, so that `ppm+p3e` and `pqm+p5e` always go together, and there's not a separate edge option necessary. 

I don't have strong feelings either way in terms of the `order` vs `method` question. As you say, I think `config_vert_remap_order` would work well, and is probably more consistent with the way other MPAS options are set up.